polaris/polaris/management/commands/process_pending_deposits.py

```diff
@@ -729,14 +866,185 @@ async def save_as_pending_signatures(cls, transaction, server):
             transaction.envelope_xdr = envelope.to_xdr()
             transaction.pending_signatures = True
             transaction.status = Transaction.STATUS.pending_anchor
-        transaction.pending_execution_attempt = False
         await sync_to_async(transaction.save)()
         await maybe_make_callback_async(transaction)
 
     @classmethod
     def handle_error(cls, transaction, message):
         transaction.status_message = message
         transaction.status = Transaction.STATUS.error
-        transaction.pending_execution_attempt = False
         transaction.save()
         logger.error(message)
+
+    @classmethod
+    def handle_transaction_submission_exception(cls, transaction, exception, message):
+        if type(exception) is TransactionSubmissionBlocked:
+            transaction.submission_status = Transaction.SUBMISSION_STATUS.blocked
+        elif type(exception) is TransactionSubmissionFailed:
+            transaction.status = Transaction.STATUS.error
+            transaction.submission_status = Transaction.SUBMISSION_STATUS.failed
+        elif type(exception) is TransactionSubmissionPending:
+            transaction.submission_status = Transaction.SUBMISSION_STATUS.pending
+        transaction.status_message = message
+        transaction.save()
+
+    @classmethod
+    def update_heartbeat(cls, key):
+        with django.db.transaction.atomic():
+            heartbeat = PolarisHeartbeat.objects.filter(key=key)[0]
+            heartbeat.last_heartbeat = datetime.datetime.now(datetime.timezone.utc)
+            heartbeat.save()
+        return
+
+    @classmethod
+    async def heartbeat_task(cls, key, heartbeat_interval):
+        """
+        Task that updates the given key at a specified interval (heatbeat_interval)
+        """
+        logger.info("heartbeat_task started...")
+        while True:
+            await sync_to_async(ProcessPendingDeposits.update_heartbeat)(key)
+            await asyncio.sleep(heartbeat_interval)
+
+    @classmethod
+    def acquire_lock(cls, key: str, heartbeat_interval: int):
+        """
+        This function creates a key in the database table 'polaris_polarisheartbeat'
+        to ensure only one instance of the calling process runs at a given time. The key 
+        is deleted when the process exists gracefully. A 'heartbeat' is utilized in the event
+        that the process does not exit gracefully. If the heartbeat's 'last_updated' value is
+        5x longer than the typical heartbeat interval, it can be assumed that the previous
+        process that created the key has crashed and a new lock can be acquired
+        """
+        attempt = 1
+        while True:
+            logger.info(
+                f"attempting to acquire lock on key: {key}, attempt #{attempt}..."
+            )
+            with django.db.transaction.atomic():
+                heartbeat, created = PolarisHeartbeat.objects.get_or_create(key=key)
```

Comment:
Ah I see, thats a good point, but I think there is a better way to write this set of queries.

What if instead of `get_or_create()`, we used `.select_for_update()` and then `.create()` if it doesn't exist, all in the same `atomic()` block?

That way, its very clear reading the code that we're locking the heartbeat key and updating the headbeat value. The way the queries are written right now don't clearly communicate that as the intention.

Reply:
i've added some additional comments in the code but i think this is exactly what get_or_create() was meant for. let me know if you feel strongly otherwise